### ISSUE TITLE
Support ECDSA subject keys

### ISSUE THREAD

**Issue by pde** [certbot/certbot#2163]
Boulder now supports [ECDSA subject keys](https://github.com/letsencrypt/boulder/pull/1298), and although it isn't deployed just yet it will be soon.  So we may want to add acme and cli support for making P384 and P521 subject keys.

**Comment by maschs** [comment]
Chrome has already dropped support for P521 and there is an open discussion to do the same in NSS/Firefox (https://bugzilla.mozilla.org/show_bug.cgi?id=1128792). 

Do P521 keys remain a viable choice without the those two supporting it?

**Comment by pde** [comment]
Sorry, I was misreading the source there.  Looks like P256 and P384 are enabled, but let me double check with the boulder team.

**Comment by kuba** [comment]
0.3.0, which is "Due by January 22, 2016" (~9 days from now) is a bit of wishful thinking, don't you think? There will likely be many compatibility problems - I'm not even sure if most recent versions of `cryptography` support it (@alex would know better)

**Comment by alex** [comment]
Sure, cryptography has supported ECDSA signatures since 0.5: https://cryptography.io/en/latest/hazmat/primitives/asymmetric/ec/

Should I file a second ticket about ECDSA support for account keys?

**Comment by pde** [comment]
@kuba you may be right that 0.3.0 is too soon; I thought there was a chance that this might just be a small change in subject key generation...

**Comment by pde** [comment]
@alex the account key change is slightly further down the road, but yes.

**Comment by kuba** [comment]
As far as I can tell no change to `acme` is necessary. `acme.client.Client.request_issuance` accepts a CSR and is oblivious to the corresponding keys type. Change must happen to the core. Removing `acme` label.

**Comment by kuba** [comment]
Thanks for info, @alex. Filed #2209.

**Comment by osirisinferi** [comment]
I've made a stub for ECDSA subject keys: #2660 

All comments are welcome, Python isn't my language at all to be honest :stuck_out_tongue:

**Comment by J0WI** [comment]
Isn't this a dupe of #2625?

**Comment by osirisinferi** [comment]
Probably more the other way around :wink:

**Comment by bmw** [comment]
Just had a conversation with a bunch of us around the office about how we want to save ECDSA keys under `/etc/letsencrypt`. The proposal we came up with is the following:

Allow users to have a single lineage that contains both ECDSA and RSA keys. This lineage will store all 8 symlinks (`cert.pem`, `privkey.pem`,...,`cert-ecdsa.pem`, `privkey-ecdsa.pem`,...) under `/etc/letsencrypt/live/lineagename`. Modifications to the lineage affect both the ECDSA and RSA certificates.

We're not planning to make changes to this directory atomic (they aren't right now either). The way things currently work is files are written to `/etc/letsencrypt/archive` first and only after that is complete are the symlinks in `/etc/letsencrypt/live` updated. We then have recovery logic that fix up the symlinks as needed on the next client run. This process can be fixed up to work with lineages with both key types for the time being.

We will also have to store more information in the renewal configuration file. For example, we currently don't save the domains in the lineage, treating the current certificate as the ground truth and reading the domains from that. This process doesn't work well if there are multiple certificates in the lineage because the names between the certs can differ if the client crashes. Such information will have to be saved in the renewal configuration file. When this file is updated with respect to the symlinks in the `live` directory is an implementation detail that has to be figured out.

Other things that were discussed include:
* Having the order they key types are requested by the user affect which certificate is obtained/renewed first by Certbot.
* Reusing certificates in the `archive` directory if we crash after obtaining one certificate.
* Reorganizing the `/etc/letsencrypt` directory to allow for more atomicity between updates.

I think all of these things are nice to have, but not essential to initially support ECDSA keys. I believe @schoen plans to work on implementing some of this in the next couple months so we can merge ECDSA support into the client relatively soon.

**Comment by ageis** [comment]
@bmw I dig it, but if you or @schoen need more inspiration then [what I said](https://github.com/certbot/certbot/issues/3349#issuecomment-325174114  about how  how Cloudflare offers every single customer [a "certificate pack"](https://support.cloudflare.com/hc/en-us/articles/216532638-FAQ-Custom-Certificate-Packs-SHA-2-and-SHA-1-) containing several types.

**Comment by bmw** [comment]
There's progress on this in https://github.com/certbot/certbot/tree/osiris-ecdsa.

**Comment by lnicola** [comment]
Yes there is.

**Comment by sabretus** [comment]
It looks like this issue is referenced to a dead branch which will never be merged, is there any way to go around this? ECDSA support is something that would be really useful

**Comment by bmw** [comment]
I'm curious. With ECDSA subject key support, who here would configure their server to use dual ECDSA/RSA certificates and who would use a single ECDSA certificate.

React with 🚀 for just an ECDSA certificate and with :eyes: for a dual certificates.

**Comment by bmw** [comment]
The design for how we'd like this to work is described in https://docs.google.com/document/d/18pr4F4SVim-bSR6izNV5ZYwolnvitTvmP6gqElDdZPU/edit?usp=sharing. The "Initial support" section describes the behavior and features we think are important to have when adding initial support for ECDSA subject keys.

**Comment by rhymeswithmogul** [comment]
> The design for how we'd like this to work is described in https://docs.google.com/document/d/18pr4F4SVim-bSR6izNV5ZYwolnvitTvmP6gqElDdZPU/edit?usp=sharing. The "Initial support" section describes the behavior and features we think are important to have when adding initial support for ECDSA subject keys.

That’s a good design. It provides new ECDSA-specific options and should ensure a smooth transition from RSA for most users.

Following those principles should allow for adding on more key types in the future. There are already some rumblings about using Edwards curves for certificate keys in the future, and the last thing Certbot developers would want is a second major refactoring for a third key type if/when EdDSA certificates get an RFC and the CA/B Forum’s blessing.

**Comment by udf2457** [comment]
I really don't understand why it is taking so long to get ECDSA support into certbot.

It is even more frustrating because certbot has this buddy-buddy relationship with Lets Encrypt where certbot is the only recommended client (https://letsencrypt.org/docs/client-options/).

If certbot wants to retain its position as exclusive recommended client then it should be leading the way by fully supporting the full range of Lets Encrypt features.  Otherwise you are just forcing people to use other client implementations, which makes the official recomendation of certbot rather pointless.

**Comment by ageis** [comment]
@bmw I would use dual ECDSA+RSA. I am an atypical user however... Basically I used to work at Cloudflare, and their free universal SSL works that way, by generating keys of all types. So I am familiar with the fact that this is even possible, but I think most aren't.

**Comment by mnordhoff** [comment]
Cloudflare's _free_ certificates are ECDSA only. The paid certificates are ECDSA + RSA.

**Comment by udf2457** [comment]
To follow up on the comment made by @ageis, popular software like `nginx` has had dual ECDSA+RSA support for some time now and is intelligent enough that you give it both certs and it uses the correct one for the client based on the client's capabilities.

**Comment by osirisinferi** [comment]
@udf2457 You're free to submit a perfectly working pull request of your ECDSA-enabled certbot of course. Complaining here actually doesn't really help or make it quicker. The small team of certbot engineers isn't sipping coffee all day laughing at users like you not having the features they want.

Also, if you really want/need ECDSA support in certbot, you're free to use my ECDSA-implementation: https://github.com/osirisinferi/certbot/tree/ecdsa-v2

**Comment by udf2457** [comment]
@osirisinferi May I point out @pde opened this request in January 2016.  Yes, that's right, four years ago. It does not take four years to implement ECDSA, no matter how small the team of engineers.

Certbot must be the only (or one of the very few) clients on the Let's Encrypt list that doesn't support ECDSA.

**Comment by bmw** [comment]
@udf2457, you made multiple comments on different Certbot issues over the weekend, many of which were overly hostile or did nothing but berate the Certbot team for not prioritizing the issues you personally care about as highly as you would like.

Please keep your comments civil, respectful, and inline with our [code of conduct](https://www.eff.org/pages/eppcode) if you are going to continue posting comments to our repository.

**Comment by adferrand** [comment]
OK @udf2457, it is the third post in a raw where you explain your frustration about the slow integration of the feature you expect. I think we got it. 

I would like to emphase a little about what happens in software engineering for a program to reach its target and gain a significant user community, like certbot with its millions of certificate issued. Maybe you forgot it or do not know about it. Everything is a question of stability, and compatibility: for a given set of functionalities, most of the work is about ensuring that nothing break, while trying to make the program work decently on the widest range of systems. And adding a new feature is increasingly more difficult as the program is run by more users. 

It is easy to add a lot of things when you have few users, the impact of regressions is low. It is nothing in comparison for a program like certbot, and yes, it is a daunting task for the small team in Certbot, when for other program of a similar popularity you have often a company and hundred of engineers to back it.

Here we are on countless hours to support all Apache versions, Nginx, all non EOL systems with some of them absurdly ancient, packaging systems with all their different procedures, global shifts of the ACME protocol, security breaches... 

So please be constructive, submit PRs if you have the time to improve certbot. And come on, what a timing, we just finally defined a proper concensus for ECDSA...

**Comment by klausenbusk** [comment]
Does using ECDSA automatic chain to the ISRG Root X2 ECDSA root?

**Comment by osirisinferi** [comment]
> Does using ECDSA automatic chain to the ISRG Root X2 ECDSA root?

Yes, in fact, **only** to X2. See the [Chain of Trust](https://letsencrypt.org/certificates/) documentation page, it has a very informative schematic of all the possible chains.
